### PR TITLE
Make a special kernel for take function

### PR DESCRIPTION
**Pull request by unnonouno**
`cupy.take` is slow as it uses `ndarray::operator(int)`. To improve its performance, I made a special kernel for a specific case such that `a.ndim==2` and `axis==0`, that is used in `F.EmbedID`.

Here is the performance comparison:

original:
```
% time python take_test.py
python take_test.py  3.70s user 1.75s system 87% cpu 6.232 total
```
nvprof
```
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.88%  4.49121s      1000  4.4912ms  4.4110ms  5.4444ms  cupy_take
```

improved:
```
% time python take_test.py
python take_test.py  0.83s user 1.07s system 71% cpu 2.657 total
```
nvprof
```
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.47%  964.29ms      1000  964.29us  904.14us  1.3731ms  cupy_take_2dim_0axis
```

And, this is the test script:
```
% cat take_test.py
import cupy

w = cupy.zeros((100, 300), dtype=cupy.float32)
ind = cupy.random.randint(0, 100, (300, 200)).astype(cupy.int32)

for e in range(1000):
    w.take(ind, axis=0)
```

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1702,7 +1712,10 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
     rdim = internal.prod(rshape)
     indices = indices.reshape(
         (1,) * len(lshape) + indices.shape + (1,) * len(rshape))
-    return _take_kernel(a, indices, cdim, rdim, out)
+    if a.ndim == 2 and axis == 0:
+        return _take_kernel_2dim_0axis(a, indices, rdim, out)
+    else:
+        return _take_kernel(a, indices, cdim, rdim, out)
```
Comment (okuta) [inline cupy/core/core.pyx]: How about using `a.reduced_view()`?

Comment (unnonouno) [inline cupy/core/core.pyx]: Do you mean `return _take_kernel(a.reduced_view(), indices, cdim, rdim, out)`?

**Comment by unnonouno** [comment]
I found that `long long` is too slow.

**Comment by unnonouno** [comment]
I found a critical bug. Please check #852 before this pr.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**